[decision] internal cache (round 3) (#4677)
ElektraInitiative/libelektra: doc/decisions/3_decided/internal_cache.md

Upon returning from `kdbGet`/`kdbSet`, set the keyname of parentKey to the key that actually is parent of the data that was loaded.
I.e. to the mountpoint of the backend that contains parentKey.
`parentKey` is already an inout-type argument, since we use both it's value and metadata to return some information.

> This behaviour was found very unexpected by @markus2330 and @atmaxinger.

Author: I've removed all personal notes now.

A partial solution would be to copy the keys from `backendData->keys`, so they are actually there, and we don't just assume they are there.
Still some extra steps are required to make this work:

```c
TEST_F (Simple, NothingToDo2)
{
	using namespace kdb;
	KDB kdb;
	KeySet ks;

	auto parentKey = "system:" + testRoot;

	EXPECT_EQ (kdb.get (ks, parentKey), 0) << "nothing to do in get";

	ks.append (Key (parentKey + "a", KEY_VALUE, "x", KEY_END));
	kdb.set (ks, parentKey);

	EXPECT_EQ (kdb.get (ks, parentKey), 0) << "nothing to do in get";
	EXPECT_TRUE (ks.lookup (parentKey + "a")) << "key a not found";
	EXPECT_EQ (ks.lookup (parentKey + "a").get<std::string> (), "x") << "key a with wrong value";
	// TODO: adding the line below breaks the test
	//	 This is because the Key instance is shared between `ks` and the internal data of `kdb`.
	// ks.lookup(parentKey + "a").set("y");

	KeySet ks2;
	EXPECT_EQ (kdb.get (ks2, parentKey), 0) << "nothing to do in get";
	EXPECT_TRUE (ks2.lookup (parentKey + "a")) << "key a not found";
	EXPECT_EQ (ks2.lookup (parentKey + "a").get<std::string> (), "x") << "key a with wrong value";
}
```

> A very simple way to make it work would be to make the keys returned by kdbGet read-only.

Author: This should be a whole decision on its own. 
There might be usecases where a read-only key name wouldn't work.
Also, how do we define "initial creation" if `keyAddName` is still allowed?

> A very simple way to make it work would be to make the keys returned by kdbGet read-only.

Reviewer: > E.g. for the Key name, I am quite in favor to make it immutable after initial creation (which might consist of several `keyAddName`).

See #4683 for a way of decoupling the keyname operations from the key API. However, I think we should absolutely avoid the word "immutable", because that is something impossible in C unless you mean something like `mprotect`.

> This should be a whole decision on its own.

Definitely, and not a high priority one, if we don't use it here.

The cache will be used to do change tracking.
From this keyset, we use `ksBelow` to return the correct keyset.

Disadvantage: mmap implementation for Windows would be needed.

The cache should be updated at the end of every kdbGet that was a cache miss.
So during kdbSet (assuming there is no external modification, i.e. conflict) the on-disk data of the cache should always be up-to-date.
The idea would be to just read the cached keyset from disk and diff against the current keyset in kdbSet.

This would mean enabling change tracking also enables the cache (or at least updating the cache, we don't have to use it in kdbGet).
If that's not wanted or if the cache data cannot be used directly for some other reason, the same approach could still be used.
We'd just have to write the keyset to disk during kdbGet and use it during kdbSet.
Since disk space is far less precious than RAM, we could even create separate files for every parent key.
If we do go down this route, kdbClose should clean up the files created by this KDB instance to avoid wasting disk space.

> This approach wouldn't be very performant (since it uses disk IO), but especially if we can use the cache data, it should be pretty easy to do.

Reviewer 2: > This is a problem of trying to record the entire discussion in the decision file. 

No, this is a problem of our discussion style. The focus should be on getting a good decision document, not to show off with amazing arguments.

IMHO the main argument is the portability, what about:

```suggestion
Disadvantages:
- It probably will not work on non-POSIX systems, as e.g. [MapViewOfFile](http://msdn.microsoft.com/en-us/library/aa366761%28VS.85%29.aspx) seems to be unreliable and it would be a huge effort to implement that.
- This approach might have a negative effect performance, because additional disk IO might occur.

Advantages:
- If data from the global cache can be reused, it might be easier to implement than the other considered alternatives.
```

> This approach wouldn't be very performant (since it uses disk IO), but especially if we can use the cache data, it should be pretty easy to do.

Reviewer: > > This is a problem of trying to record the entire discussion in the decision file.
> 
> No, this is a problem of our discussion style. The focus should be on getting a good decision document, not to show off with amazing arguments.

If _you_ are trying to show off your amazing arguments, than that _is_ a problem yes. But that's definitely not what I'm doing. I also wasn't trying to blame anybody. I was simply stating a fact. Summarising an organic discussion is not easy. Sometimes you loose context or forget to add extra remarks.

> IMHO the main argument is the portability, what about:

That was not at all my argument. My argument was very simple. The main idea is to store the data **on disk**. This by definition involves disk IO, which _could_ negatively affect performance (if not cached by the OS).

> It probably will not work on non-POSIX systems, as e.g. [MapViewOfFile](http://msdn.microsoft.com/en-us/library/aa366761%28VS.85%29.aspx) seems to be unreliable and it would be a huge effort to implement that.

Please expand on that. How exactly have you come to the conclusion it is unreliable? I would doubt that all memory mapping on Windows is unreliable. Maybe there is a different API, but memory mapped files definitely exist on Windows.

Also I found https://github.com/alitrack/mman-win32 (sadly without a license), so using the Win32 APIs to replace `mmap` seems fairly easy too.

One could argue that the problem right now is that such "broken" symlinks are not prevented.
The proposed solution doesn't completely fix the problem either

For example:

- `system:/foo` is a mountpoint, there are no other mountpoints
- `spec:/foo/bar` refers to `system:/abc` via `meta:/override`

Now in a plugin that is part of the mountpoint `system:/foo` doing a `ksLookupByName (ks, "/foo/bar", 0)` would NOT use `system:/abc`, because that key is never part of the keyset passed to this plugin.

However, in an application that used `system:/foo` as the parent, the `meta:/override` would work with your proposal.
This seems like very confusing behavior, because both the application and the plugin seemingly use the same parent key.

### Data restrictions

Make all the keys returned by kdbGet completely read-only.
To change the data it is required to append an entirely new key to replace the existing one.
Then we just need to keep a shallow copy internally.

### API restrictions

Change the API and remove KeySet from kdbGet and kdbSet also option 4 in [the operation sequences decision](../0_drafts/operation_sequences.md).

> If the keyset is owned by the KDB handle, it should not be as big surprise, if there is extra data in there.

Reviewer: ```suggestion
If the keyset is owned by the KDB handle, it should not be such a big surprise, if there is extra data in there.
```

## Decision

> Implement the copy-on-write approach.

Author: Then what's the point in splitting it into two decisions?
My point here is to say we do copy-on-write, but how the copy-on-write is implemented will be decided in the copy-on-write decision.

> Implement the copy-on-write approach.

Reviewer 2: I mean to move the text that is directly above (section above) down here. Unfortunately, in GitHub you cannot do such suggestions.

> Implement the copy-on-write approach.

Author: I've updated the decision file.

We keep a copy of all keys returned by the backends in memory.
We use `ksBelow` to only return a copy-on-write copy of keys the user requested on `kdbGet`.

In `kdbSet` we use the user-provided `KeySet` for all backends strictly below `parentKey` as before.
For the backend that contains `parentKey`, we start with the internally cached data.
We then remove everything that is at or below `parentKey` (via `ksCut`) and replace it with the data from the user-provided `KeySet`.
Keys not at or below `parentKey` therefore remain untouched.

## Rationale

Semantics can be provided without additional code or overhead in the core.
As we need copy-on-write for efficient change tracking anyway, it makes sense to also use this approach for the internal cache.

> This also does not require any changes or restrictions to the current API.

Author: I've updated it to say the solution.

## Implications

> - We need to implement the copy-on-write decision

Author: Then what's the point in splitting it into two decisions?
My point here is to say we do copy-on-write, but how the copy-on-write is implemented will be decided in the copy-on-write decision.

> - We need to implement the copy-on-write decision

Author: This is actually described in the Decision section, but I can add it here too.

> - We need to implement the copy-on-write decision

Reviewer 2: Don't duplicate it, just say something like:

```suggestion
- Before we can implement this decision, we need to implement the copy-on-write decision.
```

Please also put links if possible.

> - We need to implement the copy-on-write decision

Author: Done.

## Decision

Implement the copy-on-write approach.
As we need COW for change tracking anyway, it makes sense to also use this approach for the internal cache.
This also does not require any changes or restrictions to the current API.

We keep a copy of all keys returned by the backends in memory.
We use `ksBelow` to only return the keys the user requested on `kdbGet`.

> On `kdbSet` we use the cached in-memory keys to fill in the missing keys outside of `parentKey` that are needed for all the backends to save.

Reviewer: I think much of this is actually "Rationale".

Also there should be more details here or above. It is not really clear which keys are COW, how they are kept, etc.

I assume, the only thing we change is that instead of returning keys from `backendData->keys` directly, we return a COW copy. So the only change would be in `backendsMerge` (possibly we need two versions or a flag parameter).

I would also say things differently for `kdbSet`. I'd say we take the internal cache as a base and add the user provided `KeySet` (below `parentKey`) on top thereby replacing any overlap.

Author: > I'd say we take the internal cache as a base and add the user provided KeySet (below parentKey) on top thereby replacing any overlap.

No, then the user could never delete keys. We can only use keys outside of `parentKey` from the cache.

Reviewer: Ah that's true of course... I'd still not say "fill in the missing keys", IMO it sounds a bit weird and it's not really clear what is happening (e.g. which keys are "missing keys"). I made a concrete suggestion in another comment.

> On `kdbSet` we use the cached in-memory keys to fill in the missing keys outside of `parentKey` that are needed for all the backends to save.

Author: Great idea!
